Better tests

costas-basdekis opened the pull request:
### Description

Add a couple of useful test features
* Ability to filter by test name: `make tests-single-local-docker TEST_TARGET=foo ONLY_TEST='Name of test'`
* Ability to check that compilation fails with the wrong configuration: `exec_test_failing $1 $2 "Title"`
  * Also ability to check that specific strings are present: `exec_test_failing $1 $2 "Title" "Error that must be present" "Another expected error message"`
* Unit tests via `platformio` :tada: See [documentation](https://github.com/costas-basdekis/Marlin/blob/affcf92441582e7a2e6fe4bfa75656d19b50653f/test/README.md#unit-tests) for explanation, and some [sample tests](https://github.com/costas-basdekis/Marlin/tree/affcf92441582e7a2e6fe4bfa75656d19b50653f/Marlin/tests)
  * Auto-collect tests
  * Run unit tests via PlatformIO custom actions, after a suggestion by Ivan https://github.com/MarlinFirmware/Marlin/issues/19347

One [example](https://github.com/costas-basdekis/Marlin/blob/db9844bd9160e542d9ad3db7a9ac513d7f555b78/buildroot/tests/BIGTREE_GTR_V1_0-tests#L29-L52) of the second part:
```bash
restore_configs
...
opt_set TEMP_SENSOR_1 1
opt_set TEMP_SENSOR_2 1
opt_set TEMP_SENSOR_3 1
# We need to check that an error occurs for the following
#opt_set TEMP_SENSOR_4 1
#opt_set TEMP_SENSOR_5 1
#opt_set TEMP_SENSOR_6 1
#opt_set TEMP_SENSOR_7 1
...
strings=("TEMP_SENSOR_4 is required with 5 or more HOTENDS"
"TEMP_SENSOR_5 is required with 6 or more HOTENDS"
"TEMP_SENSOR_6 is required with 7 or more HOTENDS"
"TEMP_SENSOR_7 is required with 8 or more HOTENDS")
exec_test_failing $1 $2 "Test missing temperature sensors" "$3" "${strings[@]}"
```

### Benefits

Hopefully the ability to write better and more tests, and test not only the happy path, but also unhappy paths.

costas-basdekis committed:
Add exec_tests_failing, to test unhappy paths

By using exec_test_failing, we can test that if we omit some
configuration, or if we mismatch it, it will be caught.

You can even test that certain strings appear in the output, and if not,
the tests fail. To do that pass them as extra arguments.

costas-basdekis committed:
Add unhappy path test for missing TEMP_SENSOR_#

Also normalise error strings for that error

sjasonsmith commented:
> * Ability to filter by test name: `make tests-single-local-docker TEST_TARGET=foo ONLY_TEST='Name of test'`

I would use this. I often comment out all but one test, or run all the commands manually on the command line to replicate it.
We would need to give tests simpler names to make selecting them easy.

> * Ability to check that compilation fails with the wrong configuration: `exec_test_failing $1 $2 "Title"`
>   * Also ability to check that specific strings are present: `exec_test_failing $1 $2 "Title" "Error that must be present" "Another expected error message"`

I think this would be useful, but I worry about the time it would take to run (and write) lots of these. I don't think we would want to add them for everything, but for some key areas where build-time error detection is very important. Examples of code I would test with this:
- STM32 timer conflict detection
- STM32F1 TMC Serial port selection (to avoid dangerous hangs in Maple)
- TMC Address conflict detection

costas-basdekis commented:
@sjasonsmith 
* `ONLY_TEST`: I think we can make this be a regex, so that you can use a subset of the name (eg `8 extruders`), and also match multiple tests (eg `8 extruders|Greek LCD`)
* I think that failing compilation finishes faster, right? And the overhead of checking should be very little.
* I agree, we don't have to test absolutely everything, and can target some key areas first

costas-basdekis committed:
Mark tests as executable

To avoid the issue of git changes

costas-basdekis committed:
Add ability to run unit tests

costas-basdekis committed:
Add some first unit tests

costas-basdekis committed:
Fix `~Timer` if it hadn't initialised

costas-basdekis committed:
Amend `restore_configs` to reset all pins files

It would only do it for ramps/RAMPS, but that's an implicit assumption
that it's the only board we will ever change.

costas-basdekis committed:
Add test documentation

costas-basdekis commented:
`ONLY_TEST` is now:
* A regex
* Or a 1-based index

sjasonsmith commented:
I have been experimenting with the unit tests as I debug https://github.com/MarlinFirmware/Marlin/issues/20122.

It has been really helpful to help me understand how the ABL code works. I have now produced many passing tests, along with some that fail as well (revealing the bug!).

I have a couple thoughts about the process:
1. I would love to be able to run them easily inside VS Code. I suppose I could do it from a terminal.
Custom targets might help with this (https://github.com/MarlinFirmware/Marlin/issues/19347)
2. I need to be able to easily debug unit tests. Right now I am using a mixed Windows/WSL environment, and don't have everything properly configured. This may not be as much of an issue once I properly have my VS Code properly configured to use WSL as a "remote" machine.
3. The `RUN_TEST` macro is very easy to forget. Frameworks like googletest have macros which declare and register their test for execution at the same time, so you can't simply forget to register a test. PIO may not have anything like this, but it stands out as a weakness.

Here's a sample of what I've been doing (don't pay too much attention to names, this is experimental right now).
![image](https://user-images.githubusercontent.com/20053467/99143484-af76d700-2612-11eb-908e-785c6f763a7a.png)

costas-basdekis commented:
That's great @sjasonsmith !

1. I'll read up on that. Can you provide an example on what would VS Code require?
2. Why wouldn't you be able to debug tests right now? Since I'm quite green with C++, can you give some examples of debuggable C++ code?
3. I would love to fix that, but I don't know enough macro magic (and in my mind it looks like it's impossible) to auto-register targets. The only way I thought it would be possible was using an external pre-processing script. I'll try to read up on `googletest`

costas-basdekis commented:
So, if I understand, you mean that we should put the configuration parts (`opt_set ...`, etc) into a script, so that we can run `pio run -t test_foo`, and so we won't need a shell script to create the test `Configuration*.h`?

costas-basdekis commented:
About `googletest` here's an explanation of how to use it: https://stackoverflow.com/a/34776737

Which is pretty cool. After reading the above I can see that we could perhaps do something like this:
```c++
DECLARE_ALL_TESTS();

DECLARE_TEST(test_foo_does_bar, {
  int value = 0 ;
  ...
  TEST_ASSERT_EQUAL(0, value);
});

void test_foo() {
  RUN_ALL_TESTS();
}
```

I would have to figure out how to exactly implement the macros, but it sounds like it's possible

costas-basdekis committed:
Use macros to auto-collect code tests

costas-basdekis committed:
Use platformio custom actions to run unit tests

costas-basdekis commented:
@sjasonsmith Now we [auto collect tests](https://github.com/costas-basdekis/Marlin/commit/400cc2112daa161e5703304bc6701c4f7c5c566c)

We also use PlatformIO [custom targets](https://docs.platformio.org/en/latest/projectconf/advanced_scripting.html#custom-targets) so you can use `platformio run -t test-marlin` (all suites) or `platformio run -t marlin_test_default_config` (a single suite)
